From 6b2828340210123a4d1cfc9c7f8d995ca38b1373 Mon Sep 17 00:00:00 2001
From: Ugo Yu <ugoyu@google.com>
Date: Mon, 17 Sep 2018 15:59:30 +0800
Subject: [PATCH 1/2] DO NOT MERGE - Check SDU lower bound before allocate
 p_data

Bug: 112321180
Test: SL4A BleCocTest:test_coc_insecured_connection_write_ascii
Change-Id: Id0c9aa2097f0b6bdc2bb9fa9086daa9452188e1d
(cherry picked from commit 6fc96f847be808a4f38eae45b5e9bbc3f18b9a2d)
---
 stack/l2cap/l2c_fcr.c | 7 +++++++
 1 file changed, 7 insertions(+)

diff --git a/stack/l2cap/l2c_fcr.c b/stack/l2cap/l2c_fcr.c
index 282b171..2b041f1 100644
--- a/stack/l2cap/l2c_fcr.c
+++ b/stack/l2cap/l2c_fcr.c
@@ -849,6 +849,13 @@ void l2c_lcc_proc_pdu(tL2C_CCB *p_ccb, BT_HDR *p_buf)
             return;
         }
 
+        if (sdu_length < p_buf->len) {
+            L2CAP_TRACE_ERROR("%s: Invalid sdu_length: %d", __func__, sdu_length);
+            android_errorWriteWithInfoLog(0x534e4554, "112321180", -1, NULL, 0);
+            /* Discard the buffer */
+            osi_free(p_buf);
+            return;
+        }
 
         if ((p_data = (BT_HDR *) osi_malloc(L2CAP_MAX_BUF_SIZE)) == NULL)
         {
-- 
2.7.4

